apps/asap-server/src/handlers/webhooks/webhook-calendar-created.ts

```diff
@@ -0,0 +1,175 @@
+          } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error(error);
+          }
+        }
+      }
+
+      if (['CalendarsCreated', 'CalendarsUpdated'].includes(event)) {
+        try {
+          const resourceId = await subscribe(payload.data.id.iv, payload.id);
+
+          await calendarController.update(payload.id, {
+            resourceId,
+          });
+        } catch (error) {
+          return {
+            statusCode: 400,
+            payload: {
+              message: error.message,
+            },
+          };
+        }
+
+        return {
+          statusCode: 200,
+          payload: payload.data.id,
+        };
+      }
+
+      return { statusCode: 204 };
+    },
+  );
+
+export const subscribeToEventChangesFactory = (
+  getJWTCredentials: GetJWTCredentials,
+) => async (calendarId: string, subscriptionId: string): Promise<string> => {
+  const creds = await getJWTCredentials();
+  const client = auth.fromJSON(creds) as JWT;
+
+  client.scopes = [
+    'https://www.googleapis.com/auth/calendar',
+    'https://www.googleapis.com/auth/calendar.events',
+  ];
+  const url = `${googleApiUrl}calendar/v3/calendars/${calendarId}/events/watch`;
+  const data = {
+    id: subscriptionId,
+    type: 'web_hook',
+    address: `${asapApiUrl}/webhook/events`,
+    params: {
+      // 30 days, which is a maximum TTL
+      ttl: 2592000,
+    },
+  };
+
+  const response = await client.request<{ resourceId: string }>({
+    url,
+    method: 'POST',
+    data,
+  });
+
+  // eslint-disable-next-line no-console
+  console.log(JSON.stringify(response, null, 2));
+
+  return response.data.resourceId;
+};
+
+export type SubscribeToEventChanges = ReturnType<
+  typeof subscribeToEventChangesFactory
+>;
+
+export const unsubscribeFromEventChangesFactory = (
+  getJWTCredentials: GetJWTCredentials,
+) => async (resourceId: string, channelId: string): Promise<void> => {
+  const creds = await getJWTCredentials();
+  const client = auth.fromJSON(creds) as JWT;
+
+  client.scopes = [
+    'https://www.googleapis.com/auth/calendar',
+    'https://www.googleapis.com/auth/calendar.events',
+  ];
+  const url = `${googleApiUrl}calendar/v3/channels/stop`;
+  const data = {
+    id: channelId,
+    resourceId,
+  };
+
+  const response = await client.request({ url, method: 'POST', data });
+
+  // eslint-disable-next-line no-console
+  console.log(JSON.stringify(response, null, 2));
+};
+
+export type UnsubscribeFromEventChanges = ReturnType<
+  typeof unsubscribeFromEventChangesFactory
+>;
+
+const getJWTCredentials: GetJWTCredentials = async () => {
+  const client = new AWS.SecretsManager({ region });
+
+  const secret = await client
+    .getSecretValue({ SecretId: googleApiCredentialsSecretId })
+    .promise();
+
+  if (!('SecretString' in secret) || !secret.SecretString) {
+    throw new Error('Invalid credentials');
+  }
+
+  return JSON.parse(secret.SecretString);
```

Comment:
Maybe we should wrap this parse in a try catch just to be safe. What do you think?

Reply:
well actually i think this has to be a valid json, i dont think it would work otherwise in AWS. And even if it does work and parsing fails we will still know that it was due to the parsing error, so it kinda works no?

Plus we are not really testing this anyway

Reply:
The first part is true, but we don't want to be throwing 500s to the FE. I would prefer if we handle the errors. This would be a 502 most likely with some proper logging for debugging if needed.

Reply:
oh we do handle them, this test:
https://github.com/yldio/asap-hub/blob/3c2354b72dba45b071c4bc72e0a72e2e7ff921fa/apps/asap-server/test/handlers/webhooks/webhook-calendar-created.test.ts#L81
shows that when the subscription function rejects with any sort of error it is caught and 502 is returned, plus the error message is attached to the response. 

Unless maybe you would like me to catch and rethrow with some more meaningful message than `Unexpected token  in JSON at position...`?

Reply:
That's right, I was referring to malformed JSON parsing. That would be the only way (I see) to make the parser blow up. Your point to the aws secret manager source is valid. Otherwise it wouldn't be there in the first place.

But still, if we get a fkd up JSON, it will return a 500 because you are calling it here https://github.com/yldio/asap-hub/pull/646/files#diff-ddce2dd2a843e8f1975ec6fe560934ba7a4727d0ac379c0371a81ca171db49abR175 and that factory has not `try/catch` wrapper.

It's up to you to follow the optimistic or pessimistic approach 😉

Reply:
actually...
https://github.com/yldio/asap-hub/blob/62188a506db39c4515f5eea1b806e19be9dccb0a/apps/asap-server/src/handlers/webhooks/webhook-calendar-created.ts#L176
is where the function is passed. It is invoked here:
https://github.com/yldio/asap-hub/blob/62188a506db39c4515f5eea1b806e19be9dccb0a/apps/asap-server/src/handlers/webhooks/webhook-calendar-created.ts#L134
and here
https://github.com/yldio/asap-hub/blob/62188a506db39c4515f5eea1b806e19be9dccb0a/apps/asap-server/src/handlers/webhooks/webhook-calendar-created.ts#L98

Both of which are within try and catch, see:
https://github.com/yldio/asap-hub/blob/62188a506db39c4515f5eea1b806e19be9dccb0a/apps/asap-server/src/handlers/webhooks/webhook-calendar-created.ts#L61
and
https://github.com/yldio/asap-hub/blob/62188a506db39c4515f5eea1b806e19be9dccb0a/apps/asap-server/src/handlers/webhooks/webhook-calendar-created.ts#L70

Reply:
You are totally right! ✅ It was my bad, I thought we were running the function when passing it as an argument.